File: VB/WinWebSolution.Module/Documents.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl

Namespace WinWebSolution.Module
	<DefaultClassOptions, System.ComponentModel.DisplayName("Documents"), DefaultProperty("Name")> _
	Public MustInherit Class DocumentBase
		Inherits Note
		Protected Const AssignedToContactCriteriaBase As String = "Active=True"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private nameCore As String
		Public Property Name() As String
			Get
				Return nameCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Name", nameCore, value)
			End Set
		End Property
		Private createdByCore As ContactBase
		<DataSourceCriteria(AssignedToContactCriteriaBase)> _
		Public Property CreatedBy() As ContactBase
			Get
				Return createdByCore
			End Get
			Set(ByVal value As ContactBase)
				SetPropertyValue("CreatedBy", createdByCore, value)
			End Set
		End Property
		Private modifiedByCore As ContactBase
		<DataSourceProperty("ModifiedByCustomLookupDataSource")> _
		Public Property ModifiedBy() As ContactBase
			Get
				Return modifiedByCore
			End Get
			Set(ByVal value As ContactBase)
				SetPropertyValue("ModifiedBy", modifiedByCore, value)
			End Set
		End Property
		Private modifiedByCustomLookupDataSourceCore As XPCollection(Of ContactBase)
		Protected ReadOnly Property ModifiedByCustomLookupDataSource() As XPCollection(Of ContactBase)
			Get
				If modifiedByCustomLookupDataSourceCore Is Nothing Then
					modifiedByCustomLookupDataSourceCore = New XPCollection(Of ContactBase)(Session)
				End If
				RefreshModifiedByCustomLookupDataSource()
				Return modifiedByCustomLookupDataSourceCore
			End Get
		End Property
		Protected Overridable Sub RefreshModifiedByCustomLookupDataSource()
			modifiedByCustomLookupDataSourceCore.Criteria = ModifiedByContactCriteria
		End Sub
		Protected MustOverride ReadOnly Property ModifiedByContactCriteria() As CriteriaOperator
	End Class
	Public Class DocumentType1
		Inherits DocumentBase
		Protected Const AssignedToContactCriteria1 As String = "ObjectType.TypeName='WinWebSolution.Module.ContactType1'"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private documentType1PropertyCore As String
		Public Property DocumentType1Property() As String
			Get
				Return documentType1PropertyCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("DocumentType1Property", documentType1PropertyCore, value)
			End Set
		End Property
		<DataSourceCriteria(AssignedToContactCriteriaBase & " AND " & AssignedToContactCriteria1)> _
		Public Shadows Property CreatedBy() As ContactBase
			Get
				Return MyBase.CreatedBy
			End Get
			Set(ByVal value As ContactBase)
				MyBase.CreatedBy = value
			End Set
		End Property
		Protected Overrides ReadOnly Property ModifiedByContactCriteria() As CriteriaOperator
			Get
				Return CriteriaOperator.Parse(String.Format("{0} AND {1}", AssignedToContactCriteriaBase, AssignedToContactCriteria1))
			End Get
		End Property
	End Class
	Public Class DocumentType2
		Inherits DocumentBase
		Protected Const AssignedToContactCriteria2 As String = "ObjectType.TypeName='WinWebSolution.Module.ContactType2'"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private documentType2PropertyCore As String
		Public Property DocumentType2Property() As String
			Get
				Return documentType2PropertyCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("DocumentType2Property", documentType2PropertyCore, value)
			End Set
		End Property
		<DataSourceCriteria(AssignedToContactCriteriaBase & " AND " & AssignedToContactCriteria2)> _
		Public Shadows Property CreatedBy() As ContactBase
			Get
				Return MyBase.CreatedBy
			End Get
			Set(ByVal value As ContactBase)
				MyBase.CreatedBy = value
			End Set
		End Property
		Protected Overrides ReadOnly Property ModifiedByContactCriteria() As CriteriaOperator
			Get
				Return CriteriaOperator.Parse(String.Format("{0} AND {1}", AssignedToContactCriteriaBase, AssignedToContactCriteria2))
			End Get
		End Property
	End Class
End Namespace

File: CS/WinWebSolution.Module/Documents.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;

namespace WinWebSolution.Module {
    [DefaultClassOptions]
    [System.ComponentModel.DisplayName("Documents")]
    [DefaultProperty("Name")]
    public abstract class DocumentBase : Note {
        protected const string AssignedToContactCriteriaBase = "Active=True";
        public DocumentBase(Session session) : base(session) { }
        private string nameCore;
        public string Name {
            get { return nameCore; }
            set { SetPropertyValue("Name", ref nameCore, value); }
        }
        private ContactBase createdByCore;
        [DataSourceCriteria(AssignedToContactCriteriaBase)]
        public ContactBase CreatedBy {
            get { return createdByCore; }
            set { SetPropertyValue("CreatedBy", ref createdByCore, value); }
        }
        private ContactBase modifiedByCore;
        [DataSourceProperty("ModifiedByCustomLookupDataSource")]
        public ContactBase ModifiedBy {
            get { return modifiedByCore; }
            set { SetPropertyValue("ModifiedBy", ref modifiedByCore, value); }
        }
        private XPCollection<ContactBase> modifiedByCustomLookupDataSourceCore;
        protected XPCollection<ContactBase> ModifiedByCustomLookupDataSource {
            get {
                if (modifiedByCustomLookupDataSourceCore == null) {
                    modifiedByCustomLookupDataSourceCore = new XPCollection<ContactBase>(Session);
                }
                RefreshModifiedByCustomLookupDataSource();
                return modifiedByCustomLookupDataSourceCore;
            }
        }
        protected virtual void RefreshModifiedByCustomLookupDataSource() {
            modifiedByCustomLookupDataSourceCore.Criteria = ModifiedByContactCriteria;
        }
        protected abstract CriteriaOperator ModifiedByContactCriteria { get;}
    }
    public class DocumentType1 : DocumentBase {
        protected const string AssignedToContactCriteria1 = "ObjectType.TypeName='WinWebSolution.Module.ContactType1'";
        public DocumentType1(Session session) : base(session) { }
        private string documentType1PropertyCore;
        public string DocumentType1Property {
            get { return documentType1PropertyCore; }
            set { SetPropertyValue("DocumentType1Property", ref documentType1PropertyCore, value); }
        }
        [DataSourceCriteria(AssignedToContactCriteriaBase + " AND " + AssignedToContactCriteria1)]
        public new ContactBase CreatedBy {
            get { return base.CreatedBy; }
            set { base.CreatedBy = value; }
        }
        protected override CriteriaOperator ModifiedByContactCriteria {
            get { return CriteriaOperator.Parse(String.Format("{0} AND {1}", AssignedToContactCriteriaBase, AssignedToContactCriteria1)); }
        }
    }
    public class DocumentType2 : DocumentBase {
        protected const string AssignedToContactCriteria2 = "ObjectType.TypeName='WinWebSolution.Module.ContactType2'";
        public DocumentType2(Session session) : base(session) { }
        private string documentType2PropertyCore;
        public string DocumentType2Property {
            get { return documentType2PropertyCore; }
            set { SetPropertyValue("DocumentType2Property", ref documentType2PropertyCore, value); }
        }
        [DataSourceCriteria(AssignedToContactCriteriaBase + " AND " + AssignedToContactCriteria2)]
        public new ContactBase CreatedBy {
            get { return base.CreatedBy; }
            set { base.CreatedBy = value; }
        }
        protected override CriteriaOperator ModifiedByContactCriteria {
            get { return CriteriaOperator.Parse(String.Format("{0} AND {1}", AssignedToContactCriteriaBase, AssignedToContactCriteria2)); }
        }
    }
}